workspaces/javascript-leetcode-month/problems/2727-is-object-empty/solution.md

```diff
@@ -0,0 +1,411 @@
+
+```ts
+function isEmpty(
+  obj: Readonly<Record<PropertyKey, unknown>> | readonly unknown[],
+): boolean {
+  return Array.isArray(obj) ? obj.length === 0 : JSON.stringify(obj) === "{}";
+}
+```
+
+We could also check if stringifying gives us a length of 2, since if the object properties is not empty, that will push the length beyond 2:
+
+[View submission on LeetCode](https://leetcode.com/problems/is-object-empty/submissions/1375719502/)
+
+```ts
+function isEmpty(
+  obj: Readonly<Record<PropertyKey, unknown>> | readonly unknown[],
+): boolean {
+  return Array.isArray(obj)
+    ? obj.length === 0
+    : JSON.stringify(obj).length === 2;
+}
+```
+
+Empty arrays also get stringified to a string of length 2 (specifically `"[]"`) so we can golf once again:
+
+[View submission on LeetCode](https://leetcode.com/problems/is-object-empty/submissions/1375719967/)
+
+```ts
+const isEmpty = (o: unknown) => JSON.stringify(o).length < 3;
+```
+
+This would break in the case of something like `isEmpty(12)` but LeetCode promised us that the input will be an array or an object, so we don't have to worry about it.
+
+We should, however, feel a little bad that we're traversing entire arrays. Let's atone by writing a more efficient solution.
+
+### Using `for...in`
+
+The most careful solutions will distinguish between arrays and non-array objects, as well as check property ownership. In pure JavaScript:
+
+[View submission on LeetCode](https://leetcode.com/problems/is-object-empty/submissions/1375720258/)
+
+```js
+/**
+ * @param {Object|Array} obj
+ * @return {boolean}
+ */
+function isEmpty(obj) {
+  if (Array.isArray(obj)) {
+    return obj.length === 0;
+  }
+
+  for (const property in obj) {
+    if (Object.hasOwn(obj, property)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+```
+
+If you like ternaries as much as I do, you may put the loop in a helper function. Note that :
+
+[View submission on LeetCode](https://leetcode.com/problems/is-object-empty/submissions/1375728095/)
+
+```ts
+function isEmptyObject(obj: Readonly<Record<string, unknown>>): boolean {
+  for (const property in obj) {
+    if (Object.hasOwn(obj, property)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+function isEmpty(
+  obj: Readonly<Record<PropertyKey, unknown>> | readonly unknown[],
+): boolean {
+  return Array.isArray(obj)
+    ? obj.length === 0
+    : isEmptyObject(obj as Readonly<Record<PropertyKey, unknown>>);
+}
+```
+
+If we don't care to be careful, we can skip the array-ness and ownership checks. LeetCode accepts this:
+
+[View submission on LeetCode](https://leetcode.com/problems/is-object-empty/submissions/1375734860/)
+
+```ts
+function isEmpty(
+  obj: Readonly<Record<PropertyKey, unknown>> | readonly unknown[],
+): boolean {
+  for (const _property in obj) {
+    return false;
+  }
+
+  return true;
+}
+```
+
+## Answers to Bonus Questions
+
+1. What's the difference between `obj instanceof Array` and `Array.isArray(obj)`?
```

Comment:
This is new to me and really useful, when you're dealing with embedded web pages!